crates/subspace-farmer/src/single_disk_farm/plot_cache.rs

```diff
@@ -168,21 +170,31 @@ impl DiskPlotCache {
 
         // Make sure offset is after anything that is already plotted
         if element_offset < plotted_bytes {
-            // Remove entry since it was overridden with a sector already
+            // Remove entry since it was overwritten with a sector already
             self.cached_pieces.write().map.remove(key);
             MaybePieceStoredResult::No
         } else {
             MaybePieceStoredResult::Yes
         }
     }
 
-    /// Store piece in cache if there is free space, otherwise `Ok(false)` is returned
+    /// Store piece in cache if there is free space, and return `Ok(true)`.
+    /// Returns `Ok(false)` if there is no free space, or the farm or process is shutting down.
     pub(crate) async fn try_store_piece(
         &self,
         piece_index: PieceIndex,
         piece: &Piece,
     ) -> Result<bool, DiskPlotCacheError> {
         let offset = {
+            // First, do a quick concurrent check for free space with a read lock.
+            let cached_pieces = self.cached_pieces.read();
+            if cached_pieces.next_offset.is_none() {
+                return Ok(false);
+            };
+            // Release the read lock, to avoid potential deadlocks.
+            drop(cached_pieces);
```

Comment:
Forgot to push this fix yesterday, taking a read and write lock at the same time is an immediate deadlock (which hangs a test).